src/main/java/com/microsoft/azure/relay/ClientWebSocket.java

```diff
@@ -200,60 +201,60 @@ boolean isOpen() {
+		return writeAsync(data, timeout, true, WriteMode.BINARY);
 	}
 	
 	/**
 	 * Sends the data to the remote endpoint within a timeout in one of the WriteModes.
 	 * 
 	 * @param data Message to be sent.
 	 * @param timeout The timeout to connect to send the data within. May be null to indicate no timeout limit.
+	 * @param isEnd Indicates if the data sent is the end of a message
 	 * @param mode The type of the message to be sent.
 	 * @return A CompletableFuture which completes when websocket finishes sending the bytes.
 	 * @throws TimeoutException Throws when the sending task does not complete within the given timeout.
 	 */
-	CompletableFuture<Void> writeAsync(Object data, Duration timeout, WriteMode mode) {
+	CompletableFuture<Void> writeAsync(Object data, Duration timeout, boolean isEnd, WriteMode mode) {
 		if (this.isOpen()) {
 			if (data == null) {
 				// TODO: Log warns sending nothing because message is null
 				return CompletableFuture.completedFuture(null);
 			}
 			else {
-				return CompletableFutureUtil.timedRunAsync(timeout, () -> {
-					RemoteEndpoint.Basic remote = this.session.getBasicRemote();
-					
-					try {
-						if (mode.equals(WriteMode.TEXT)) {
-							String text = data.toString();
-							RelayLogger.logEvent("writingBytes", this, "text");
-							Writer writer = remote.getSendWriter();
-							writer.write(text);
-							writer.close();
-							RelayLogger.logEvent("doneWritingBytes", this, String.valueOf(text.length()));
-						}
-						else {
-							ByteBuffer bytes = null;
-							
-							RelayLogger.logEvent("writingBytes", this, "binary");
-							if (data instanceof byte[]) {
-								bytes = ByteBuffer.wrap((byte[]) data);
-							} 
-							else if (data instanceof ByteBuffer) {
-								bytes = (ByteBuffer) data;
+				RemoteEndpoint.Basic remote = this.session.getBasicRemote();
+				RelayLogger.logEvent("writingBytes", this, mode.toString());
+				
+				// The websocket API will throw if multiple sends are attempted on the same websocket simultaneously
+				// Therefore a lock is used on the send operation
+				return this.socketLock.acquireThenCompose(timeout, () -> {
+					return CompletableFutureUtil.timedRunAsync(timeout, () -> {
+						try {
+							if (mode.equals(WriteMode.TEXT)) {
+								String text = data.toString();
+								remote.sendText(text, isEnd);
+								RelayLogger.logEvent("doneWritingBytes", this, String.valueOf(text.length()));
 							}
-							else if (data instanceof String){
-								bytes = ByteBuffer.wrap(data.toString().getBytes(StringUtil.UTF8));
+							else {
+								ByteBuffer bytes = null;
+								if (data instanceof byte[]) {
+									bytes = ByteBuffer.wrap((byte[]) data);
+								} 
+								else if (data instanceof ByteBuffer) {
+									bytes = (ByteBuffer) data;
+								}
+								else if (data instanceof String){
+									bytes = ByteBuffer.wrap(data.toString().getBytes(StringUtil.UTF8));
+								}
```

Comment:
Isn't it possible for an Object to not fall into any of these if/else-if combinations? Ideally this API should throw instead of doing an empty send in that case.